linera-core/src/join_set_ext.rs

```diff
@@ -5,113 +5,136 @@
 
-#[cfg(not(web))]
-impl JoinSetExt for JoinSet<()> {
-    fn spawn_task<F>(&mut self, future: F) -> TaskHandle<F::Output>
-    where
-        F: Future + Send + 'static,
-        F::Output: Send,
-    {
-        let (output_sender, output_receiver) = oneshot::channel();
-
-        let abort_handle = self.spawn(async move {
-            let _ = output_sender.send(future.await);
-        });
-
-        TaskHandle {
-            output_receiver,
-            abort_handle,
-        }
-    }
+    /// An extension trait for the [`JoinSet`] type.
+    pub trait JoinSetExt: Sized {
+        /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn_local`].
+        ///
+        /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+        /// [`AbortHandle`] to cancel execution of the task.
+        fn spawn_task<F: Future + 'static>(&mut self, future: F) -> TaskHandle<F::Output>;
 
-    async fn await_all_tasks(&mut self) {
-        while self.join_next().await.is_some() {}
-    }
+        /// Awaits all tasks spawned in this [`JoinSet`].
+        fn await_all_tasks(&mut self) -> impl Future<Output = ()>;
 
-    fn reap_finished_tasks(&mut self) {
-        while self.try_join_next().is_some() {}
+        /// Reaps tasks that have finished.
+        fn reap_finished_tasks(&mut self);
     }
-}
 
-#[cfg(web)]
-impl JoinSetExt for JoinSet<()> {
-    fn spawn_task<F>(&mut self, future: F) -> TaskHandle<F::Output>
-    where
-        F: Future + 'static,
-    {
-        let (output_sender, output_receiver) = oneshot::channel();
-
-        let abort_handle = self.spawn_local(async move {
-            let _ = output_sender.send(future.await);
-        });
-
-        TaskHandle {
-            output_receiver,
-            abort_handle,
+    impl JoinSetExt for JoinSet {
+        fn spawn_task<F: Future + 'static>(&mut self, future: F) -> TaskHandle<F::Output> {
+            let (abort_handle, abort_registration) = AbortHandle::new_pair();
+            let (send_done, recv_done) = oneshot::channel();
+            let (send_output, recv_output) = oneshot::channel();
+            let future = async move {
+                let _ = send_output.send(future.await);
+                let _ = send_done.send(());
+            };
+            self.0.push(recv_done);
+            wasm_bindgen_futures::spawn_local(
+                future::Abortable::new(future, abort_registration).map(drop),
+            );
+
+            TaskHandle {
+                output_receiver: recv_output,
+                abort_handle,
+            }
+        }
+
+        async fn await_all_tasks(&mut self) {
+            stream::iter(&mut self.0)
+                .then(|x| x)
+                .map(drop)
+                .collect()
+                .await
+        }
+
+        fn reap_finished_tasks(&mut self) {
+            for mut done in self.0.drain(..) {
+                while done.try_recv() == Ok(None) {}
+            }
```

Comment:
What's the difference between this and `self.0.clear()`, effectively?

Reply:
This waits for all the futures to complete before returning, whereas dropping them via `clear` would return immediately while they're still running (and eventually cause a `Canceled` error in the future, though we do ignore that anyway).

Reply:
Ah, right, I misread that. It's waiting very busily!

Reply:
It is, but I think this is meant to be a catch for an error case, not something that happens in normal execution.